lib/fs.js

```diff
@@ -376,52 +263,30 @@ function readFileAfterStat(err, stats) {
   context.read();
 }
 
-function readFileAfterRead(err, bytesRead) {
-  var context = this.context;
-
-  if (err)
-    return context.close(err);
-
-  if (bytesRead === 0)
-    return context.close();
-
-  context.pos += bytesRead;
-
-  if (context.size !== 0) {
-    if (context.pos === context.size)
-      context.close();
-    else
-      context.read();
-  } else {
-    // unknown size, just read until we don't get bytes.
-    context.buffers.push(context.buffer.slice(0, bytesRead));
-    context.read();
-  }
-}
-
-function readFileAfterClose(err) {
-  var context = this.context;
-  var buffer = null;
-  var callback = context.callback;
-
-  if (context.err || err)
-    return callback(context.err || err);
+function readFile(path, options, callback) {
+  callback = maybeCallback(callback || options);
+  options = getOptions(options, { flag: 'r' });
+  if (!ReadFileContext)
+    ReadFileContext = require('internal/fs/readfilecontext');
+  const context = new ReadFileContext(callback, options.encoding);
+  context.isUserFd = isFd(path); // file descriptor ownership
+  const req = new FSReqWrap();
+  req.context = context;
+  req.oncomplete = readFileAfterOpen;
 
-  try {
-    if (context.size === 0)
-      buffer = Buffer.concat(context.buffers, context.pos);
-    else if (context.pos < context.size)
-      buffer = context.buffer.slice(0, context.pos);
-    else
-      buffer = context.buffer;
-
-    if (context.encoding)
-      buffer = buffer.toString(context.encoding);
-  } catch (err) {
-    return callback(err);
+  if (context.isUserFd) {
+    process.nextTick(function tick() {
+      req.oncomplete(null, path);
+    });
+    return;
   }
 
-  callback(null, buffer);
+  path = getPathFromURL(path);
```

Comment:
Can you validate the path before instantiating the req wrap?

Reply:
Actually, no, not currently. Note the condition immediately before this, moving the validatePath up would be a breaking behavior change without some other refactoring in this method that is best left to a separate more focused PR.

Reply:
What I will do for sure, tho, is open that separate PR ;-)